walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,61 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    // Given a set of existing events, a new request should be arranged at right time
+    // the restriction is that the attendees in the new request should be present at one meeting at a time
+    // this function returns a collection of appropriate time for the new request
+    Collection<String> requestAttendees = request.getAttendees();
+    List<TimeRange> res = new LinkedList<>();
+    res.add(TimeRange.WHOLE_DAY);
+    for (Event event : events) {
+      Set<String> Attendees = event.getAttendees();
+      boolean flag = false;
+      for (String attendee : Attendees) {
+        if (requestAttendees.contains(attendee)) {
+          flag = true;
+          break;
+        }
+      }
+      if (!flag) {
+        continue;
+      }
+      TimeRange badTime = event.getWhen();
+      ListIterator it = res.listIterator();
+      while (it.hasNext()) {
+        TimeRange goodTime = (TimeRange) it.next();
+
+        if (!goodTime.overlaps(badTime)) {
+          continue;
+        }
+        if (goodTime.contains(badTime)) {
+          it.remove();
+          it.add(TimeRange.fromStartDuration(goodTime.start(), badTime.start() - goodTime.start()));
+          it.add(TimeRange.fromStartDuration(badTime.end(), goodTime.end() - badTime.end()));
+        } else if (badTime.contains(goodTime)) {
```

Comment:
I think you'd better check this first, then when badTime is exactly the same as goodTime, you won't add two empty time range to the set.

Reply:
> I think you'd better check this first, then when badTime is exactly the same as goodTime, you won't add two empty time range to the set.

If I do so, there is still problem like goodTime and badTime have the same start but different end. So I just added two "if" check in my code.